### ISSUE TITLE
BUG: `package.json:files` field breaks `git` installs.

### ISSUE THREAD

**Issue by pbrianmackey** [FormidableLabs/radium#581]
When I run `npm install radium --save` it errors:

> ENOENT.  This is most likely not a problem with npm itself and is related to npm not being able to 
> find a file.  Error no -4058

Looking at the attached log I saw it quit because of something found in node_modules.  So I deleted everything in node_modules and tried again.  Same error

[npm-debug.txt](https://github.com/FormidableLabs/radium/files/123667/npm-debug.txt)

**Comment by pbrianmackey** [comment]
I went through the same process on a Mac and it worked.  My windows machine can't install.

**Comment by tuckerconnelly** [comment]
Wasn't able to install 0.16.6 on Mac. Looks like 0.16.5 worked though :)

**Comment by ianobermiller** [comment]
cc @tptee who released 0.16.6

**Comment by tptee** [comment]
I can't reproduce this issue.

Tried `npm install --save radium` on the following:
- OSX 10.11, node 5.6.0, npm 3.7.0
- Windows 10, node 4 LTS, npm 2
- Windows 10, node 5.6.0, npm 3

All succeeded. Best advice I can give is to check your node and npm versions and to run a `npm cache clean`.

**Comment by ryan-roemer** [comment]
Wow, that is a weird ENOENT:

```
162751 error enoent ENOENT, rename 'C:\Code\Git\ReactStylingRadium\node_modules\radium\node_modules\babel-preset-stage-1\node_modules\babel-preset-stage-2\node_modules\babel-preset-stage-3\node_modules\babel-plugin-transform-async-to-generator\node_modules\babel-helper-remap-async-to-generator\node_modules\babel-traverse\node_modules\babel-code-frame\node_modules\chalk\node_modules\strip-ansi\node_modules\ansi-regex\package.json.988ca2f102d18e457afb9bcbc99b3671'
```

Does that _look_ like a correct path for any babel-familiar folks here?

**Comment by ryan-roemer** [comment]
You know, it could that windows max path thing. I heard npm 3 does a much better job with this.

@pbrianmackey -- Can you try updating to `npm install -g npm@3` and if possible update node to `4.3`+?

**Comment by davidgilbertson** [comment]
I've had two strange issues.
Node 4.4
NPM 3 (so not the max path thing for me)
Windows 10
I was installing some other package (babel stage-2 for memory) and it actually _removed_ the radium install. The command I ran had _nothing to do_ with radium. But radium showed up in the error and wasn't in `node_modules` after.

Also, I just deployed to google app engine and got an error about radium.

Sorry I have no logs, if it happens again I will report back with more details.

**Comment by davidgilbertson** [comment]
I'm back with an actual error. I'm getting this when trying to push my project to Google App Engine.
```
Step 1 : FROM gcr.io/google_appengine/nodejs
 ---> 3b4050d21a95
Step 2 : RUN /usr/local/bin/install_node '^4.2.0'
 ---> Running in e6c6cf82e61f
 ---> 2419708367af
Removing intermediate container e6c6cf82e61f
Step 3 : COPY . /app/
 ---> b30ea5ecaed5
Removing intermediate container aebc4b8d8582
Step 4 : RUN npm install --unsafe-perm ||   ((if [ -f npm-debug.log ]; then       cat npm-debug.log;     fi) && false)
 ---> Running in c7be6a481534

> radium@0.17.1 postinstall /app/node_modules/radium
> cd lib || npm run lib
ERROR: gcloud crashed (error): [Errno 10054] An existing connection was forcibly closed by the remote host
```
It's using:
```
node v4.4.3
npm  v3.8.7
```
And `radium ^0.17.1`

It runs just fine (and is great) locally.

**Comment by davidjbeveridge** [comment]
Another variant:

I can `npm install --save radium@0.17.1` just fine. But if I use the git repo, it blows up:

```
$ npm install --save 'https://github.com/FormidableLabs/radium.git#v0.17.1'

> radium@0.17.1 postinstall /my/project/node_modules/radium
> cd lib || npm run lib

sh: line 0: cd: lib: No such file or directory

> radium@0.17.1 lib /my/project/node_modules/radium
> npm run babel && rimraf lib/__tests__ lib/__mocks__


> radium@0.17.1 babel /my/project/node_modules/radium
> rimraf lib && babel src/ -d lib/

src/ doesn't exist

npm ERR! Darwin 14.5.0
npm ERR! argv "/Users/davidjbeveridge/.nvm/versions/node/v5.11.0/bin/node" "/my/project/node_modules/.bin/npm" "run" "babel"
npm ERR! node v5.11.0
npm ERR! npm  v3.9.5
npm ERR! code ELIFECYCLE
npm ERR! radium@0.17.1 babel: `rimraf lib && babel src/ -d lib/`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the radium@0.17.1 babel script 'rimraf lib && babel src/ -d lib/'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the radium package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     rimraf lib && babel src/ -d lib/
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs radium
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls radium
npm ERR! There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! Please include the following file with any support request:
npm ERR!     /my/project/node_modules/radium/npm-debug.log

npm ERR! Darwin 14.5.0
npm ERR! argv "/Users/davidjbeveridge/.nvm/versions/node/v5.11.0/bin/node" "/my/project/node_modules/.bin/npm" "run" "lib"
npm ERR! node v5.11.0
npm ERR! npm  v3.9.5
npm ERR! code ELIFECYCLE
npm ERR! radium@0.17.1 lib: `npm run babel && rimraf lib/__tests__ lib/__mocks__`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the radium@0.17.1 lib script 'npm run babel && rimraf lib/__tests__ lib/__mocks__'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the radium package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run babel && rimraf lib/__tests__ lib/__mocks__
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs radium
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls radium
npm ERR! There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! Please include the following file with any support request:
npm ERR!     /my/project/node_modules/radium/npm-debug.log
npm ERR! Darwin 14.5.0
npm ERR! argv "/Users/me/.nvm/versions/node/v5.11.0/bin/node" "/Users/davidjbeveridge/.nvm/versions/node/v5.11.0/bin/npm" "install" "--save" "https://github.com/FormidableLabs/radium.git#v0.17.1"
npm ERR! node v5.11.0
npm ERR! npm  v3.9.6
npm ERR! code ELIFECYCLE

npm ERR! radium@0.17.1 postinstall: `cd lib || npm run lib`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the radium@0.17.1 postinstall script 'cd lib || npm run lib'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the radium package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     cd lib || npm run lib
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs radium
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls radium
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /my/project/npm-debug.log
```

I'm using:

Mac OSX 10.10.5
Node v5.11.0
NPM 3.9.6

**Comment by ryan-roemer** [comment]
@davidjbeveridge -- That's really weird because from `git` (not `npm`), `src` should definitely exist. Can you check if `/my/project/node_modules/radium/src` exists?

**Comment by davidjbeveridge** [comment]
`/my/project/node_modules/radium` doesn't exist; I'm not super clear on the npm installation process, but it looks like it's being deleted, or never linked, because of the failed postinstall step.

**Comment by davidjbeveridge** [comment]
Also, weird: if I clone the repo locally, and install via  `file:///` URI, it works just fine.

**Comment by ryan-roemer** [comment]
Wacky. Just repro'ed in totally empty directory with `npm install --save 'https://github.com/FormidableLabs/radium.git#v0.17.1'`. Investigating more...

**Comment by ryan-roemer** [comment]
Well, now this is totally perplexing. I've added a branch https://github.com/FormidableLabs/radium/compare/temp-test-git-npm-install-breakage (based off `#v0.17.1`) with some extra debug information:

```js
"lib-debug": "echo \"Running 'lib' command in: $(pwd)\" && find . -ls | egrep -v \"\\.\/(node_modules|\\.git)\"",
```

which produces:

```
$ npm install --save 'https://github.com/FormidableLabs/radium.git#temp-test-git-npm-install-breakage'

> fsevents@1.0.12 install /Users/rye/Desktop/TEMP_RADIUM/node_modules/fsevents
> node-pre-gyp install --fallback-to-build

[fsevents] Success: "/Users/rye/Desktop/TEMP_RADIUM/node_modules/fsevents/lib/binding/Release/node-v46-darwin-x64/fse.node" is installed via remote

> radium@0.17.1 postinstall /Users/rye/Desktop/TEMP_RADIUM/node_modules/radium
> cd lib || npm run lib

sh: line 0: cd: lib: No such file or directory

> radium@0.17.1 lib /Users/rye/Desktop/TEMP_RADIUM/node_modules/radium
> npm run lib-debug && npm run babel && rimraf lib/__tests__ lib/__mocks__


> radium@0.17.1 lib-debug /Users/rye/Desktop/TEMP_RADIUM/node_modules/radium
> echo "Running 'lib' command in: $(pwd)" && find . -ls | egrep -v "\./(node_modules|\.git)"

Running 'lib' command in: /Users/rye/Desktop/TEMP_RADIUM/node_modules/radium
62122348        0 drwxr-xr-x    6 rye              staff                 204 Jun  7 09:53 .
62123238       32 -rw-r--r--    1 rye              staff               12821 Jun  7 09:53 ./CHANGELOG.md
62123586        8 -rw-r--r--    1 rye              staff                1082 Jun  7 09:53 ./LICENSE.md
62124008       24 -rw-r--r--    1 rye              staff               11278 Jun  7 09:53 ./package.json
62122937       16 -rw-r--r--    1 rye              staff                5769 Jun  7 09:53 ./README.md

> radium@0.17.1 babel /Users/rye/Desktop/TEMP_RADIUM/node_modules/radium
> rimraf lib && babel src/ -d lib/

src/ doesn't exist
```

I have no idea why we're missing nearly all the repo files...

**Comment by davidjbeveridge** [comment]
Is it possible that somehow `rifraf` is taking `lib && babel src/ -d lib/` all as literal arguments instead of being interpreted by the shell? that might cause src to get deleted.

**Comment by davidjbeveridge** [comment]
Nm. That doesn't make sense. `rimraf` shouldn't be running until afterward.

**Comment by ryan-roemer** [comment]
Cuhrazy. I've not _removed_ `postinstall` and left just debugging information. It doesn't error, but at the end we are left with:

```
$ ls node_modules/radium
CHANGELOG.md LICENSE.md   README.md    package.json
```

And that's it. I think I'm either missing something, or ???

**Comment by davidjbeveridge** [comment]
Well, it's nice to know I'm not crazy :)

**Comment by ryan-roemer** [comment]
Found it. The `package.json:files` field limits what files get installed in **both** npm and git deps, which effectively breaks git installs completely (because the only files there are not in git source 😉  ).

This diff _will_ install and build just fine from git source:

```diff
diff --git a/package.json b/package.json
index e239415..d8db921 100644
--- a/package.json
+++ b/package.json
@@ -3,10 +3,6 @@
   "version": "0.17.1",
   "description": "A set of tools to manage inline styles on React elements",
   "main": "lib/index.js",
-  "files": [
-    "dist",
-    "lib"
-  ],
   "repository": {
     "type": "git",
     "url": "https://github.com/formidablelabs/radium.git"
```

Given that we've already run into some issues with `package.json:files`, I think maybe it's time to consider doing the ol' manual combination of:

* `.gitignore`
* `.npmignore`

which can effectively support both git and npm installs keeping the relevant files needed...

/cc @alexlande @ianobermiller

**Comment by ianobermiller** [comment]
Sounds fine to me. I think I used `files` because a whitelist is usually easier to maintain than a blacklist (it either works, or doesn't, but doesn't include anything extra). Funny that we've had the prepublish code to make git installs work, and they haven't been working anyway!

**Comment by jmcriffey** [comment]
@ryan-roemer @ianobermiller I actually set everything up with a .npmignore file and it seems like npm still excludes everything as if it were still using the files list in package.json. Looking at https://github.com/npm/npm/issues/7262, it seems like we would need to be inefficient in either npm or github to be installable in both places.

**Comment by alexlande** [comment]
Resolved by #730. I'll cut a new release shortly.
